### ISSUE TITLE
Cross Building failure(for AARCH64) on x86_64 Ubuntu(64bit) 

### ISSUE THREAD

**Issue by Qiupetter** [hduong85/arm_adb#7]
Hi, Hduong,
I have a problem on Cross Building (for AARCH64) on x86_64 Ubuntu(64bit) .
Could you pls give me some help?

build log:

libtool: link: g++  -fPIC -DPIC -shared -nostdlib /usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/5/crtbeginS.o  libcutils/.libs/libcutils_la-socket_inaddr_any_server_unix.o libcutils/.libs/libcutils_la-socket_local_client_unix.o libcutils/.libs/libcutils_la-socket_local_server_unix.o libcutils/.libs/libcutils_la-socket_network_client_unix.o libcutils/.libs/libcutils_la-load_file.o libcutils/.libs/libcutils_la-threads.o libcutils/.libs/libcutils_la-sockets.o   -L/usr/lib/gcc-cross/aarch64-linux-gnu/5 -L/usr/lib/gcc/x86_64-linux-gnu/5 -L/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu -L/usr/lib/gcc/x86_64-linux-gnu/5/../../../../lib -L/lib/x86_64-linux-gnu -L/lib/../lib -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib -L/usr/lib/gcc/x86_64-linux-gnu/5/../../.. -lstdc++ -lm -lc -lgcc_s /usr/lib/gcc/x86_64-linux-gnu/5/crtendS.o /usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/crtn.o  -g -O2   -Wl,-soname -Wl,libcutils.so.0 -o .libs/libcutils.so.0.0.0
/usr/bin/ld: libcutils/.libs/libcutils_la-socket_inaddr_any_server_unix.o: Relocations in generic ELF (EM: 183)
/usr/bin/ld: libcutils/.libs/libcutils_la-socket_inaddr_any_server_unix.o: Relocations in generic ELF (EM: 183)
/usr/bin/ld: libcutils/.libs/libcutils_la-socket_inaddr_any_server_unix.o: Relocations in generic ELF (EM: 183)
/usr/bin/ld: libcutils/.libs/libcutils_la-socket_inaddr_any_server_unix.o: Relocations in generic ELF (EM: 183)
/usr/bin/ld: libcutils/.libs/libcutils_la-socket_inaddr_any_server_unix.o: Relocations in generic ELF (EM: 183)
libcutils/.libs/libcutils_la-socket_inaddr_any_server_unix.o: error adding symbols: File in wrong format
collect2: error: ld returned 1 exit status
Makefile:532: recipe for target 'libcutils.la' failed
make[2]: *** [libcutils.la] Error 1
make[2]: Leaving directory '/home/wenguangqiu/arm_adb/lib'
Makefile:452: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/wenguangqiu/arm_adb'
Makefile:363: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by qhuyduong** [comment]
Hi @Qiupetter ,

Which toolchain are you using to cross-compile? Can you check if your toolchain includes both gcc and g++ (e.g: aarch64-linux-gnu-gcc and aarch64-linux-gnu-g++)? 

Thanks.

**Comment by Qiupetter** [comment]
Hi Hduong,

Thank you for your kind reply~ I only installed "aarch64-linux-gnu-gcc".....
After aarch64-linux-gnu-g++ installed, build is successful.

But there is an issue about dynamic link:
$ file adb
adb: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, BuildID[sha1]=2e0b8e1460fe41f903755e802a8b0284af05e1b3, with debug_info, not stripped

Is there any way to change to static link?

Thank you again ^^.

**Comment by qhuyduong** [comment]
Hi @Qiupetter ,

Actually I intended to make it static link (by adding -static to LDFLAGS as you can see in src/Makefile.am). So it should be static link already. You can check by just copying the adb binary to your aarch64 platform and see if it can run successfully. If it can, so the "dynamically linked" in this scenario may be different story :)

Thanks.

**Comment by Qiupetter** [comment]
Hi Hduong,

Ah, I used your released "adb" binary, so that file must be dynamic linked.
I will try the adb binary built by myself~

Thank you very much ^^.

**Comment by Qiupetter** [comment]
Hi Hduong,

I tried to build adb binary, then pushed it into android phone, but it can not run normally.
"No such file or directory.."
I think it caused by the lack of "interpreter /lib/ld-linux-aarch64.so.1" on android phone...

Thank you.

**Comment by qhuyduong** [comment]
Hi @Qiupetter ,

Well, the original idea of this arm_adb is to enable adb on the linux-based systems, not android. As you know, google provides adb binary for x86 systems only. To be able to use adb in Linux on other platforms such as arm, arm64, ..., one usually has to rebuild adb from source, and the procedure might not be straight forward. That's why I created this repository to make things easier.

Android, despite being built based on Linux, has so many differences from Linux, especially the standard libraries.  Some of basic Linux libraries don't exist in Android, that's why you see the error.

In summary, if you want to run adb on an android phone, this repo is not an option :).

Thanks,
Huy Duong

**Comment by Qiupetter** [comment]
Hi Hduong,

Thank you for your replies! I think you are right.
